torch/optim/optimizer.py

```diff
@@ -422,11 +422,16 @@ def _process_value_according_to_param_policy(param: Tensor, value: Tensor, param
                 capturable = pg["capturable"] if "capturable" in pg else False
                 break
 
-        if key != "step" or capturable or fused:
+        if key == 'step':
+            if capturable or fused:
+                return value.to(dtype=torch.float32, device=param.device)
+            else:
+                return value
+        else:
```

Comment:
I'd rewrite this for loop this way:
```
if key == 'step':
    if capturable or fused:
        return value.to(dtype=torch.float32, device=param.device)
    # otherwise, we DON'T cast step in any way.
    return value
if param.is_floating_point():
    return value.to(dtype=param.dtype, device=param.device)
return value.to(device = param.device)
```

This also means we're saying "don't touch step UNLESS it should join the parameter for fused/capturable". I'm curious if we should cast it to float in all cases.

Reply:
I feel like omitting the else blocks makes this code less clear tbh, if this is a nit, imma not fix. Let me know if you feel really strongly about this.

Reply:
The way the initialization is written the step ignores the dtype of the param, it's either fp64  by default or fp32 if capturable/fused, so this matches that behavior.

Reply:
I think the comments are more useful than the else blocks (they just make the indentation deeper).

Reply:
Yea lol I'm aware of the way it's been written, I'm more intrigued by whether we should have casted to fp32 vs fp64, but your change is fine for now. My question is beyond the scope and can be rectified later if need be.